# Fundamentals of Data Analysis
# Tasks

**Ciaran Moran**

***


### 1 - The Collatz Conjecture

https://en.wikipedia.org/wiki/Collatz_conjecture

Program to run all nubers 1 to 10000 though the Collztz Conjecture.

The mathematical formula being:-
$$ 
f(x) =  
\begin{cases}
    \\
    \ x \div 2  & if \hspace{0.1cm} x \hspace{0.1cm} is \hspace{0.1cm} even  
    \\[3ex]
    \ 3x + 1  & otherwise
\end{cases}
$$


For the above latex formula I figured it out using the following websites - 

https://linuxhint.com/use-latex-jupyter-notebook/

https://tex.stackexchange.com/questions/670525/align-math-expressions-next-to-the-bracket

https://www.overleaf.com/learn/latex/Line_breaks_and_blank_spaces#Horizontal_blank_spaces

And a direct Google for "latex division symbol"


In [ ]:
def f(x):
    # If x is even, then divide by 2
    if x % 2 == 0:
        return x // 2
    else:
        return (3 * x) + 1
    

## Collatz Function

The is where we apply the Collatz code for the mathematical function

I have printed out the value of 'x' to ensure I am testing everything from 1 to 10000 inclusive.

For this I only print if less than 10 or greater than 9990 for efficiency purposes.

In [ ]:
def collatz(x):
    print(f'>>> Testing number {x}')
    while x != 1:
        print(x, end=', ')
        x = f(x)
    print(x)


### The Loop
Next loop from 1 - 10000.

Loops start at 0, so there are a number of options for the loop.

We could loop 1 - 10001, or 0-10000 + 1 i.e. i+1.

Personally I prefer using 0-10000 + 1.


In [ ]:
for i in range(10000):
    collatz(i+1)

As observed in the output above, all numbers 1-10000 run through the Collatz function results in an ending number of 1. Therefore in this scanario the Collatz function is proved to be true.

*** 

## 2 - Penguin data set

Give an overview of the famous penguins data set,2 explaining 2 mwaskom/seaborn-data: Data repository
for seaborn examples. Aug. 30, 2023. url:
https : / / github . com / mwaskom /
seaborn - data / blob / master /
penguins.csv (visited on 08/30/2023).
the types of variables it contains. Suggest the types of variables
that should be used to model them in Python, explaining your
rationale.

**References used for understanding the Penguin Data Set**

https://www.kaggle.com/code/parulpandey/penguin-dataset-the-new-iris

https://www.robots.ox.ac.uk/~vgg/data/penguins/


In [2]:
import matplotlib.pyplot as plt 
import random
import datetime
import pandas as pd 
import seaborn as sns
import numpy as np

df = pd.read_csv('data/penguins/penguins.csv')
print("Head...")
print(df.head())
print()
print("Describe...")
df.describe



Head...
  species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0  Adelie  Torgersen            39.1           18.7              181.0   
1  Adelie  Torgersen            39.5           17.4              186.0   
2  Adelie  Torgersen            40.3           18.0              195.0   
3  Adelie  Torgersen             NaN            NaN                NaN   
4  Adelie  Torgersen            36.7           19.3              193.0   

   body_mass_g     sex  
0       3750.0    MALE  
1       3800.0  FEMALE  
2       3250.0  FEMALE  
3          NaN     NaN  
4       3450.0  FEMALE  

Describe...


<bound method NDFrame.describe of     species     island  bill_length_mm  bill_depth_mm  flipper_length_mm  \
0    Adelie  Torgersen            39.1           18.7              181.0   
1    Adelie  Torgersen            39.5           17.4              186.0   
2    Adelie  Torgersen            40.3           18.0              195.0   
3    Adelie  Torgersen             NaN            NaN                NaN   
4    Adelie  Torgersen            36.7           19.3              193.0   
..      ...        ...             ...            ...                ...   
339  Gentoo     Biscoe             NaN            NaN                NaN   
340  Gentoo     Biscoe            46.8           14.3              215.0   
341  Gentoo     Biscoe            50.4           15.7              222.0   
342  Gentoo     Biscoe            45.2           14.8              212.0   
343  Gentoo     Biscoe            49.9           16.1              213.0   

     body_mass_g     sex  
0         3750.0    MALE  

***These are the 3 types of penguins included in the data set***
1. Chinstrap
2. Gentoo
3. Adelie


![3 3 penguin types in the penguin data set](https://allisonhorst.github.io/palmerpenguins/reference/figures/lter_penguins.png)



### 2. Give an overview of the famous penguins data set(2) explaining the types of variables it contains. Suggest the types of variables that should be used to model them in Python, explaining your rationale.

Websites referenced.
 - https://www.kaggle.com/code/parulpandey/penguin-dataset-the-new-iris
 - https://www.kaggle.com/datasets/parulpandey/palmer-archipelago-antarctica-penguin-data
 - https://towardsdatascience.com/penguins-dataset-overview-iris-alternative-9453bb8c8d95
 - https://www.analyticsvidhya.com/blog/2022/04/data-exploration-and-visualisation-using-palmer-penguins-dataset/
- https://www.w3schools.com/python/python_datatypes.asp

 


The penguin data set contains 7 variables.
1. species
    - Chinstrap
    - Gentoo
    - Adelie
2. island
    - this is the particular island that the data was captured on
    - the islands are in the Palmer Archipelago in Antarctica
    - 3 islands are in the data set
        - Dream
        - Torgersen
        - Biscoe
3. bill_length_mm
    - the length of the penguin's bill, measured in mm
4. bill_depth_mm
    - the depth of the penguin's bill, measured in mm
5. flipper_length_mm
    - the length of the penguin's flipper, measured in mm
6. body_mass_g
    - the body mass of the penguin, measured in grams
7. sex
    - MALE / FEMALE

### Suggested Python data types 

| Variable | Suggested Python data type | Reason |
| ----------- | ----------- | ----------- |
| species | string | this is a text field for the species of the penguin e.g. Chinstrap |
| island | string | this is a text field for the name of the island e.g. Dream  |
| bill_length_mm | float | this column contains floating point values e.g. 39.1 |
| bill_depth_mm | float | this column contains floating point values e.g. 18.7 |
| flipper_length_mm | int | this column contains whole integer numbers e.g. 181 |
| body_mass_g | int |  this column contains whole integer numbers e.g. 3750  |
| sex | string | this is a text field for the sex of the penguin e.g. MALE / FEMALE   |






### 3. For each of the variables in the penguins data set(3) suggest what probability distribution from the numpy random distributions list is the most appropriate to model the variable.

References used
 - https://numpy.org/doc/stable/reference/random/index.html#random-sampling-numpy-random
 - https://numpy.org/doc/stable/reference/random/index.html#module-numpy.random
 - https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html
 - https://numpy.org/doc/stable/reference/routines.html
 - https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html#numpy-random-normal
 - https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html
 - https://stackoverflow.com/questions/31037298/pandas-get-column-average-mean
 - https://datascienceparichay.com/article/pandas-get-standard-deviation-of-one-or-more-column


 

For the **float** variables:-

- bill_length_mm
- bill_depth_mm


A suitable distribution would be **numpy.random.normal**


This will *random samples from a normal (Gaussian) distribution*. 
It takes in floats as parameters.

*random.normal(loc=0.0, scale=1.0, size=None)*
* loc - float or array_like of floats
* scale - float or array_like of floats
* size - int or tuple of ints, optional


In [ ]:
# Example code.
# References
# https://numpy.org/doc/stable/reference/random/generated/numpy.random.normal.html#numpy-random-normal
# https://stackoverflow.com/questions/31037298/pandas-get-column-average-mean
# https://datascienceparichay.com/article/pandas-get-standard-deviation-of-one-or-more-column


# Generate a random float between min and max for column bill_length_mm
random_bill_length_mm = np.random.normal(df.loc[:, 'bill_length_mm'].mean(), df['bill_length_mm'].std(), 1)
print("Random Float :", random_bill_length_mm[0])
print("Random Float Rounded: ",int(np.round(random_bill_length_mm*100))/100)

For the **integer** variables:-

- flipper_length_mm
- body_mass_g

A suitable distribution would be **numpy.random.randint**

This will *return random integers from the discrete uniform distribution of the specified dtype in the “half-open” interval*.

In [ ]:
# Example code.
# Generate a random integer between min and max for column flipper_length_mm
random_bill_length_mm = np.random.randint(df['flipper_length_mm'].min(), df['flipper_length_mm'].max()+1)
print("Random integer value for flipper_length_mm :", random_bill_length_mm)

For the **string** variables:-

- species
- island
- sex

A suitable distribution would be **numpy.random.choice**
Although other distributions would also work 
e.g. normal distribution numpy.random.normal or uniform distribution numpy.random.uniform

https://numpy.org/doc/stable/reference/random/generated/numpy.random.choice.html

*Parameters: 1-D array-like or int*
*If an ndarray, a random sample is generated from its elements*

In other words, supply a list of valid strings to use, and we can randomly select one of those

In [ ]:
# Example code.
# Generate a random string from distinct column string values for column species
# Referenced websites
# https://stackoverflow.com/questions/47368962/unique-strings-in-a-pandas-dataframe
species_distinct_arr = np.unique(np.concatenate(df['species'].str.split(';\s*').values))
random_species = np.random.choice(species_distinct_arr, 1)
print("Random string value for species :", random_species[0])

### 4. Suppose you are flipping two coins, each with a probability p of giving heads. Plot the entropy of the total number of heads versus p.

References Used
https://stackoverflow.com/questions/67703875/np-random-binomial-vs-random-choices-for-simulating-coin-flips
https://en.wikipedia.org/wiki/Entropy_(information_theory)#:~:text=The%20entropy%20is%20zero%3A%20each,coin%20toss%20is%20always%20certain.
https://stackoverflow.com/questions/477237/how-do-i-simulate-flip-of-biased-coin
https://stackoverflow.com/questions/21761630/using-python-to-analyse-coin-tossing-statistics
https://stackoverflow.com/questions/75497605/python-flipping-x-coins-x-times-using-loops-but-receiving-odd-output
https://towardsdatascience.com/how-to-code-a-fair-coin-flip-in-python-d54312f33da9
https://cmdlinetips.com/2018/12/simulating-coin-toss-experiment-with-binomial-random-numbers-using-numpy/
https://medium.com/geekculture/how-to-simulate-coin-flips-using-binomial-distribution-in-python-165aae7daf22
https://numpy.org/doc/stable/reference/random/generated/numpy.random.binomial.html


### 5. Create an appropriate individual plot for each of the variables in the penguin data set.

***

## End